### DISCUSSION TITLE
Memory leaks that occur when tokens are frequently updated

### DISCUSSION THREAD

**Discussion by min-sung-kwon** [OpenIDC/mod_auth_openidc#1228] [General]
Situation: The `OIDCRefreshAccessTokenBeforeExpiry` value is set to receive a refresh token once a minute.
Result: Server memory usage exceeds 80-90% every 2-3 days in version 2.4.11.3.
Action: Update to version 2.4.15.7
Progress: We thought the issue had been resolved by confirming that the increase in memory usage had decreased, but we found that the increase had only decreased but was still increasing. Currently, server memory usage exceeds 80-90% once every 2-3 weeks.

The issue is as above and I would like to resolve the issue.

**Comment by zandbelt** [comment]
I don't think mod_auth_openidc is leaking memory, our valgrind reports don't show that. How did you come to this conclusion? Using OS process memory tools may not be as reliable.

**Reply by min-sung-kwon** [reply]
I believe there is a memory leak in `mod_auth_openidc` for the following reasons:
1. A memory leak occurred after changing the `OIDCRefreshAccessTokenBeforeExpiry` setting to update the token once a minute. This setting was changed within the last month, and memory usage had never increased this much in about 2 years before the change.
2. Restart Apache and memory usage will stabilize. This is evidence that the Apache module is taking up memory, and when considered together with reason 1, it is believed that a memory leak occurs in `mod_auth_openidc`.

**Reply by min-sung-kwon** [reply]
3. In the case of methods other than `mod_auth_openidc`, which updates tokens every minute, memory leaks do not occur.

**Reply by zandbelt** [reply]
a. which tool are you using - on which platform - to measure the memory allocation?
b. please provide your mod_auth_openidc related Apache config

**Reply by min-sung-kwon** [reply]
a. We are monitoring memory usage with a self-developed tool based on salt.
b. The apache config is as follows, and sensitive information is hidden as {Secret}.

```
OIDCProviderMetadataURL {Secret}
OIDCClientID {Secret}
OIDCClientSecret {Secret}
OIDCRedirectURI {Secret}
OIDCCryptoPassphrase {Secret}

OIDCRemoteUserClaim {Secret}
OIDCAuthNHeader {Secret}
OIDCClaimPrefix

OIDCOAuthServerMetadataURL {Secret}
OIDCOAuthIntrospectionEndpointParams {Secret}={Secret}
OIDCOAuthClientID {Secret}
OIDCOAuthClientSecret {Secret}

OIDCCacheType redis
OIDCRedisCacheServer {Secret}

OIDCSessionInactivityTimeout 86400
OIDCStateMaxNumberOfCookies 3 true
OIDCStateTimeout 3600
OIDCSessionMaxDuration 0

# For reference, the access token expiration time is 28800 seconds.
OIDCRefreshAccessTokenBeforeExpiry 28740 logout_on_error
OIDCDefaultLoggedOutURL {Secret}
OIDCUnAuthAction 401 "(-n %{HTTP:Sec-Fetch-Dest} && %{HTTP:Sec-Fetch-Dest} != 'iframe' && %{HTTP:Sec-Fetch-Dest} != 'document')"

OIDCCookie {Secret}

<LocationMatch "{Secret}" >
  AuthType auth-openidc
  Require valid-user
</LocationMatch>
```

**Reply by zandbelt** [reply]
a self-developed tool doesn't provide proof, please use valgrind or other standard tools; on which OS/Linux version/platform are you?

**Reply by zandbelt** [reply]
here's some Valgrind output of 1-second interval requests with `OIDCRefreshAccessTokenBeforeExpiry 1000000`

10 runs:
> HEAP SUMMARY:
> apache-1  | ==971==     in use at exit: 10,574 bytes in 20 blocks
> apache-1  | ==971==   total heap usage: 46,565 allocs, 46,545 frees, 11,519,246 bytes allocated
> ...
> apache-1  | ==971==    still reachable: 10,574 bytes in 20 blocks
> 

100 runs:
> HEAP SUMMARY:
> apache-1  | ==1058==     in use at exit: 10,574 bytes in 20 blocks
> apache-1  | ==1058==   total heap usage: 46,565 allocs, 46,545 frees, 11,519,246 bytes allocated
> ...
> apache-1  | ==1058==    still reachable: 10,574 bytes in 20 blocks

**Reply by min-sung-kwon** [reply]
Currently, we do not have an environment to diagnose memory leaks using Valgrind, etc., so we are recognizing the phenomenon through the memory usage of CentOS7.
Are the Valgrind diagnostic logs you showed the result of testing in an environment where tokens are updated for each request?

**Reply by zandbelt** [reply]
> Currently, we do not have an environment to diagnose memory leaks using Valgrind, etc., so we are recognizing the phenomenon through the memory usage of CentOS7. 

we'll try on CentOS 7 as well then, and perhaps you can try the behavior on CentOS 9; which version of cjose are you using?

> Are the Valgrind diagnostic logs you showed the result of testing in an environment where tokens are updated for each request?

yes

**Reply by min-sung-kwon** [reply]
We are using cjose version 0.6.1.

**Reply by zandbelt** [reply]
please also try with the latest cjose from https://github.com/OpenIDC/cjose/releases/tag/v0.6.2.3 as memory leaks have been addressed in newer versions here as well; I'd be interested to learn about your findings on CentOS 9 as CentOS 7 has proven to be more problematic for us in the past (i.e. lingering semphores, wich seems to be OS related).

**Reply by min-sung-kwon** [reply]
We will update the cjose version next week and share the results. Thanks for your help.

**Reply by zandbelt** [reply]
FWIW: cjose is a long shot; our tests on CentOS 7 show no definite memory leaks in Valgrind but there are "possible leaks" that increase with the number of runs, which is what your tool may be reporting on as well; the "possible leaks" are not related to mod_auth_openidc in itself but to the old versions of libapr/libapr-util in use on CentOS 7 and OpenSSL 1.0.2; all of which should be solved by migrating to a more recent and supported OS
